### PR TITLE
[Feature] TensorDict without device

### LINKED ISSUE #394: [Feature Request] TensorDict without device
## Motivation

TensorDict have a static and defined device. This was thought as a core property of this class, until we thought about using tensordict to store the weights of models (either for saving them or to pass them to modules in functional settings). As [pointed here](https://github.com/facebookresearch/rl/pull/387#discussion_r964113227) a model could be split across devices, and how TensorDict should behave in this case is less defined.

## Solution

We could leave some space for TensorDict to be either set on a device if it is explicitly asked, or left with an undefined device otherwise.
Here are a few examples:

```python
tensordict = TensorDict({"a": torch.randn(3, 4)}, [])
tensordict.device # returns None

tensordict = TensorDict({"a": torch.randn(3, 4, device=0)}, [])
tensordict.device # returns None

tensordict = TensorDict({"a": torch.randn(3, 4)}, [], device=0)
tensordict.device # returns torch.device("cuda:0"), "a" in on "cuda:0"

tensordict = TensorDict({"a": torch.randn(3, 4, device=0)}, [], device=0)
tensordict.device # returns torch.device("cuda:0")

tensordict = TensorDict({"a": torch.randn(3, 4, device=0)}, [], device=0)
tensordict["b"] = torch.randn(3)
tensordict["b"].device  # "b" has been sent to device cuda:0

tensordict = TensorDict({"a": torch.randn(3, 4, device=0)}, [])
tensordict.device # returns None
tensordict = tensordict.to("cuda:0")
tensordict.device # returns torch.device("cuda:0")

```

At first glance, this will require changes in
- `TensorDict.to` and `TensorDictBase` subclasses `to` methods
- `TensorDict._process_tensor`, which will send the tensors to device only if the device is strictly defined
- `__init__` will not have to set the device if it is not specified
- Tests in `test/test_tensordict.py` will have to be changed accordingly
- `TensorDict.device` will fail if no device is set, whereas `_device_safe` won't. I guess we'll want `device` to behave like `_device_safe`? Or we may still want device to fail if it's not set but that may cause some pain to the users. I'm not against it though, better a good exception than a messy behvaviour (i.e. in this case all the "returns None" here above would in fact be exceptions).

### PR DESCRIPTION
**Pull request by tcbegley**
## Description

This PR enables `TensorDict` to be used without a defined device. 

- If the device is defined, all elements must life on the same device, if it is undefined, elements can live on different devices.
- If the `TensorDict` has no device, then adding elements with an explicit device will _not_ set the device on the `TensorDict`. 
- If the `TensorDict` has a defined device, then elements added to the `TensorDict` will be moved to this device.

See `test/test_tensordict.py:test_tensordict_device` for many examples of intended behaviour.

This change required a number of existing code and tests to be updated where either there was an attempt to access `tensordict.device` (which raises a `RuntimeError` if device is not set) or the test relied on the previous behaviour of the `TensorDict` device being set when an element was added.

## Motivation and Context

Closes #394 

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

Please advise on whether documentation / example updates are appropriate.

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/facebookresearch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] My change requires a change to the documentation. (Probably?)
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by tcbegley — COMMENTED** [review]
Still TODO:

- [x] Add test for formatting of `_ErrorInteceptor` messages (should give the nested hierarchy of keys at which device mismatch occured)
- [x] Add `clear_device` method to `TensorDict` as alternative to confusing `td.to(None)`
- [x] Potentially resolve failing tests: the basic tests are passing locally but I had a couple of issues when trying to run those with optional dependencies such as `gym`. Think it might have been my setup so waiting to see what happens with the CI.

**Comment by vmoens** [comment]
cc @zou3519 => addresses the TensorDict on multiple devices discussion

File: torchrl/data/tensordict/tensordict.py [hunk 1]
```diff
@@ -1890,22 +1868,20 @@ def device(self) -> torch.device:
...
-                "because device cannot be updated in-place. To update device, use "
-                "tensordict.to(new_device), which will return a new tensordict "
-                "on the new device."
-            )
+        raise RuntimeError(
+            "device cannot be set using tensordict.device = device, "
+            "because device cannot be updated in-place. To update device, use "
+            "tensordict.to(new_device), which will return a new tensordict "
+            "on the new device."
+        )
```
[the remark this answers is not in the document]
Comment (vmoens) [reply torchrl/data/tensordict/tensordict.py:1887]: Yes you'd be surprised by how many people try to do stuff like this thinking that it sends their tensors to a certain device or whatever. I like to have setter methods that tell people that what they're trying to do is not supported.

**Review by vmoens — COMMENTED** [review]
This is of the greatest quality!
Can we make _device_safe() public, and give an error when calling a non-set device that is explicit about what the difference is between the two?

**Comment by tcbegley** [comment]
I've made `device_safe` public and updated the error messages for `device` to direct people to use it.

**Review by vmoens — APPROVED** [review]
Perfect! Let's land this

**Pull request merged**